### PR TITLE
Adds coordinator (ReAct) mode for enhanced reasoning

### LINKED ISSUE #115: Adopt Iterative Reasoning (ReAct) with Subagent Delegation for Task Execution
Bernard should transition from single-pass or monolithic execution toward a compositional execution model that combines:

* Iterative reasoning and action (ReAct pattern)
* Delegation of scoped work to subagents

The primary agent (Bernard) should remain focused on high-level coordination, while subagents are used to solve narrowly defined problems in isolated contexts.

This enables improved reasoning clarity, reduced context bloat, and more reliable task execution.

---

## Problem Statement

Current limitations:

* Tasks are handled within a single, growing context window
* Reasoning and execution are tightly coupled in one agent
* Context accumulation leads to:

  * degraded reasoning quality
  * irrelevant information influencing decisions
* Complex tasks lack clear boundaries or decomposition strategy

This results in:

* Reduced accuracy on multi-step or compound tasks
* Increased hallucination due to overloaded context
* Difficulty maintaining focus on the current objective

---

## Desired Behavior

Bernard should operate as a coordinator, not the sole executor.

### 1. Iterative execution via ReAct

* Use a Reason → Act → Observe loop for task progression
* Determine next steps incrementally rather than upfront

### 2. Delegate scoped work to subagents

* Identify when a task can be isolated into a smaller, self-contained problem
* Spawn or invoke a subagent with:

  * a clearly defined objective
  * minimal required context
* Subagent returns a result that is:

  * focused
  * bounded
  * relevant to the parent task

### 3. Maintain context discipline

* Avoid accumulating unnecessary intermediate reasoning in the main agent
* Use subagents to reset context and reduce noise
* Treat subagent outputs as structured observations

### 4. Preserve high-level control in the main agent

* Bernard is responsible for:

  * task decomposition
  * sequencing
  * decision-making
* Subagents are responsible for:

  * execution of a single, well-defined task

---

## Subagent Usage Guidelines (Behavioral)

Bernard should prefer subagent delegation when:

* A task requires focused reasoning over a specific domain or dataset
* Intermediate steps risk polluting the main context
* The problem can be clearly defined with:

  * inputs
  * expected output shape

Bernard should avoid subagent usage when:

* The task is trivial or single-step
* Delegation overhead outweighs benefit
* The problem cannot be meaningfully scoped

---

## Non-Goals

This issue does not define:

* How subagents are implemented or instantiated
* Communication protocols between agents
* Specific memory or context-passing mechanisms
* Infrastructure or orchestration details

This defines behavioral expectations, not system design.

---

### A. Hierarchical execution

* Complex tasks are decomposed into smaller units
* At least some units are delegated to subagents where appropriate

### B. Context isolation

* Subtasks are executed in isolated contexts
* Main agent context remains focused on high-level reasoning

### C. Clear responsibility boundaries

* Main agent does not attempt to solve all subproblems directly
* Subagents are used for targeted execution

### D. Improved reasoning quality

* Reduced evidence of context drift or irrelevant reasoning
* Outputs are more focused and grounded

### E. Effective integration of results

* Subagent outputs are incorporated as observations
* Subsequent reasoning reflects those results accurately

---

## Risks & Considerations

* Over-delegation leading to fragmentation
* Increased latency from spawning multiple agents
* Poorly scoped subagents producing low-quality outputs
* Need for guardrails to prevent recursive or unnecessary delegation

---

## Success Metrics

* Improved accuracy on compound/multi-step tasks
* Reduction in context size growth within main agent
* Increased task completion reliability
* More consistent use of tools and delegated execution

---

## Conceptual Flow

User Request
↓
[Main Agent: Thought]
↓
[Delegate? → Yes → Subagent Task]
↓
[Subagent Execution]
↓
[Result / Observation]
↓
[Main Agent: Thought]
↓
(repeat until complete)

---

## Future Extensions (Out of Scope)

* Planner / Critic integration (PAC model)
* Subagent specialization (role-based agents)
* Tool selection optimization within subagents
* Execution tracing and observability

---

## Strategic Shift

This change moves Bernard from:

"Solve the problem"

to:

"Coordinate the solving of the problem"

That distinction enables the system to scale in complexity without collapsing under context pressure.

### PR DESCRIPTION
**Pull request by phillt**
Introduces a new coordinator mode that enables iterative reasoning with structured planning and delegation capabilities:

- **Enhanced reasoning loop**: Implements think → act → evaluate → decide pattern for more deliberate decision-making
- **Structured planning**: Adds plan management with status tracking (pending, in_progress, done, cancelled, error)
- **Improved delegation**: Encourages decomposition of work into scoped sub-tasks with better context discipline
- **Visual feedback**: Provides user-visible thoughts, evaluations, and plan updates during execution
- **Adaptive step budget**: Automatically triples iteration limit (clamped at 150) to accommodate the more thorough approach
- **Context optimization**: Caps subagent result lengths to prevent context bloat in parent agent

The mode is configurable via environment variable or REPL command and includes comprehensive test coverage and enforcement mechanisms to ensure plan completion.

closes #115

### CODE REVIEW

File: src/agent.ts [hunk 1]
```diff
@@ -178,6 +182,78 @@ In addition to stating your plan in text, persist it to scratch for reliability:
 - After any mutation (file write, git commit, API call), immediately verify the outcome with a read-only command.
 - Your work will be reviewed by a critic agent afterward. Only claim what you can prove with tool output.`;
 
+const REACT_COORDINATOR_PROMPT = `## Coordinator Mode (Active)
+
+You are operating as a coordinator, not a sole executor. Your primary role is to decompose, delegate, and synthesize — not to do all work yourself.
+
```
[the remark this answers is not in the document]
Comment (phillt) [reply src/agent.ts:188]: Good catch — confirmed wrong issue link. Updated the PR body to reference #115 (Adopt Iterative Reasoning (ReAct) with Subagent Delegation). #114 (prompt-optimizer) is unrelated and remains open.

File: src/tools/result-cap.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+/**
+ * Caps a sub-agent or specialist result string to prevent context bloat in the parent agent.
+ */
+export function capSubagentResult(
+  text: string,
+  maxChars: number = SUBAGENT_RESULT_MAX_CHARS,
+): string {
+  if (text.length <= maxChars) return text;
+  return text.slice(0, maxChars) + `\n...[output truncated at ${maxChars} chars]`;
```
[the remark this answers is not in the document]
Comment (phillt) [reply src/tools/result-cap.ts]: Good catch — fixed in the latest commit. `capSubagentResult` now reserves space for the truncation marker so the total returned length is guaranteed `<= maxChars`. Tests updated to assert `capped.length === cap` plus a new `length never exceeds maxChars` boundary test and an edge-case test for when `maxChars` is smaller than the marker itself (returns a truncated marker). See `src/tools/result-cap.ts` and `src/tools/result-cap.test.ts`.
